In [1]:
import pandas as pd

In [2]:
#nyc_taxi_2019-07

df_2019 = pd.read_csv(
    filepath_or_buffer='../../pandas-workout-data/data/nyc_taxi_2019-07.csv',
    usecols=['passenger_count', 'total_amount', 'payment_type']
    )
df_2019.head(5)

passenger_count  payment_type  total_amount
0              1.0           1.0          4.94
1              1.0           2.0         20.30
2              1.0           1.0         70.67
3              1.0           1.0         66.36
4              0.0           1.0         15.30

In [3]:
df_2019['year'] = 2019
df_2019.head(5)

passenger_count  payment_type  total_amount  year
0              1.0           1.0          4.94  2019
1              1.0           2.0         20.30  2019
2              1.0           1.0         70.67  2019
3              1.0           1.0         66.36  2019
4              0.0           1.0         15.30  2019

In [4]:
df_2020 = pd.read_csv(
    filepath_or_buffer='../../pandas-workout-data/data/nyc_taxi_2020-07.csv',
    usecols=['passenger_count', 'total_amount', 'payment_type']
    )
df_2020.head(5)

passenger_count  payment_type  total_amount
0              1.0           2.0          9.30
1              1.0           1.0         27.80
2              1.0           2.0         22.30
3              1.0           1.0         14.16
4              1.0           2.0          7.80

In [5]:
df_2020['year'] = 2020
df_2020.head(5)

passenger_count  payment_type  total_amount  year
0              1.0           2.0          9.30  2020
1              1.0           1.0         27.80  2020
2              1.0           2.0         22.30  2020
3              1.0           1.0         14.16  2020
4              1.0           2.0          7.80  2020

In [6]:
len(df_2019) + len(df_2020)

7110831

In [7]:
df = pd.concat([df_2019, df_2020], axis=0)
df

passenger_count  payment_type  total_amount  year
0                   1.0           1.0          4.94  2019
1                   1.0           2.0         20.30  2019
2                   1.0           1.0         70.67  2019
3                   1.0           1.0         66.36  2019
4                   0.0           1.0         15.30  2019
...                 ...           ...           ...   ...
800407              NaN           NaN         83.50  2020
800408              NaN           NaN         19.78  2020
800409              NaN           NaN         38.45  2020
800410              NaN           NaN         29.77  2020
800411              NaN           NaN         51.90  2020

[7110831 rows x 4 columns]

How many rides were taken in 2019 and 2020, and what is the difference between these two figures?

In [8]:
rides_2019 = len(df_2019)
rides_2020 = len(df_2020)
difference = rides_2019 - rides_2020
difference

5510007

How much money (in total) was collected in 2019 and 2020, and what was the difference between these two figures?

In [9]:
amount_2019 = df.loc[df['year'] == 2019, 'total_amount'].sum()
amount_2020 = df.loc[df['year'] == 2020, 'total_amount'].sum()
print(f"2019: ${amount_2019:.2f} vs 2020: ${amount_2020:.2f}")

2019: $123761823.33 vs 2020: $14912844.09


In [10]:
print(f"Difference: ${(amount_2019 - amount_2020):.2f}")

Difference: $108848979.24


Did the proportion of trips with more than one passenger change dramatically?

In [11]:
trips_more_than_one_passenger_2019 = df.loc[(df['passenger_count'] > 1) & (df['year'] == 2019), 'passenger_count'].count()
trips_more_than_one_passenger_2019

np.int64(1778686)

In [12]:
trips_more_than_one_passenger_2020 = df.loc[(df['passenger_count'] > 1) & (df['year'] == 2020), 'passenger_count'].count()
trips_more_than_one_passenger_2020

np.int64(152050)

In [13]:
df.loc[10] # This was to verify the values in index 10

passenger_count  payment_type  total_amount  year
10              1.0           1.0         15.95  2019
10              2.0           1.0         28.30  2020

Yes it changed dramatically

In [14]:
print(f"Percetage: {((trips_more_than_one_passenger_2020/trips_more_than_one_passenger_2019)*100):.2f}%")

Percetage: 8.55%


Did people use cash (i.e., payment_type of 2) less in 2020 than in 2019?

In [15]:
cash_2019 = df.loc[(df['year'] == 2019) & (df['payment_type'] == 2), 'payment_type'].count()
cash_2019

np.int64(1801718)

In [16]:
cash_2020 = df.loc[(df['year'] == 2020) & (df['payment_type'] == 2), 'payment_type'].count()
cash_2020

np.int64(236433)

Yes, they did

In [17]:
df.loc[df['year'] == 2020, 'payment_type'].count()

np.int64(737565)

In [18]:
# Throughout the year

In [19]:
# Did the proportion of trips with more than passenger change dramatically?
df.loc[(df['year'] == 2019) & 
       (df['passenger_count'] > 1), 'passenger_count'].count() / df.loc[df['year'] == 2019, 'payment_type'].count()

np.float64(0.2833900000955953)

In [21]:
df.loc[(df['year'] == 2020) & 
       (df['passenger_count'] > 1), 'passenger_count'].count() / df.loc[df['year'] == 2020, 'payment_type'].count()

np.float64(0.2061513222563435)

In [23]:
# Did people use cash less in 2019 or 2020?
df.loc[(df['year'] == 2019) & 
       (df['payment_type'] == 2), 'payment_type'].count() / df.loc[df['year'] == 2019, 'payment_type'].count()

np.float64(0.2870595845428793)

In [24]:
df.loc[(df['year'] == 2020) & 
       (df['payment_type'] == 2), 'payment_type'].count() / df.loc[df['year'] == 2020, 'payment_type'].count()

np.float64(0.320558865998251)

### Beyond exercise

Use the corr method on df to find the correlations among the columns. How would you interpret these results?

In [25]:
df.corr()

passenger_count  payment_type  total_amount      year
passenger_count         1.000000      0.016410      0.014943 -0.049558
payment_type            0.016410      1.000000     -0.138561  0.029277
total_amount            0.014943     -0.138561      1.000000 -0.019706
year                   -0.049558      0.029277     -0.019706  1.000000

Show, with a single command, the difference in descriptive statistics for total_amount beween 2019 and 2020. Round values to use no more than 2 digits after the decimal point.

In [26]:

(
    df.loc[df['year'] == 2020, 'total_amount'].describe().round(2) - 
    df.loc[df['year'] == 2019, 'total_amount'].describe().round(2)
)

count   -5510007.00
mean          -0.98
std           -0.75
min           53.20
25%           -0.50
50%           -0.60
75%           -0.75
max        -4672.45
Name: total_amount, dtype: float64

If we assume that zero-passenger trips are for delivering packages, how were those affected during the pandemic? Show the proportion of such trips in 2019 vs. 2020.

In [27]:
df.loc[(df['year'] == 2019) & 
       (df['passenger_count'] == 0), 'passenger_count'].count() / df['passenger_count'].count()

np.float64(0.01666432611802781)

In [28]:
df.loc[(df['year'] == 2020) & 
       (df['passenger_count'] == 0), 'passenger_count'].count() / df['passenger_count'].count()

np.float64(0.0027809994974354953)

Turns out that the proportion of zero-person taxi rides shrank as well, from about 1.6% in 2019 to 0.2% in 2020. Honestly, I expected it to be higher during the pandemic!